src/devices/src/virtio/file_traits.rs

```diff
@@ -425,14 +425,13 @@ impl FileReadWriteAtVolatile for imago::SyncFormatAccess<imago::file::File> {
             return Ok(0);
         }
 
-        let mut offset = offset as usize;
-        let mut read = 0;
-        for buf in bufs {
-            let bytes = self.read_at_volatile(*buf, offset as u64)?;
-            offset += bytes;
-            read += bytes;
-        }
-        Ok(read)
+        let (iovec, _guard) = imago::io_buffers::IoVectorMut::from_volatile_slice(bufs);
+        let full_length = iovec
+            .len()
+            .try_into()
+            .map_err(|e| Error::new(ErrorKind::InvalidData, e))?;
+        self.readv(iovec, offset)?;
+        Ok(full_length)
```

Comment:
I think this hunk (and the next one in `write_vectored_at_volatile()`) should be squashed into the previous commit.

Reply:
Yes you're right. I edited the wrong commit. Nice catch!